### PR TITLE
Feature 6: file io

### LINKED ISSUE #2: [Feature 6] File I/O
TDLR: Feature that allows a saved timetable to be exported and a timetable file to be imported as well as read a JSON file containing all courses at the University of Toronto Faculty of Arts & Science from the Fall (F) and Winter (S) sessions. 

Summary:
In Feature 6, a JSON file containing all courses at the University of Toronto Faculty of Arts & Science from the Fall (F) and Winter (S) sessions can be read and converted into different entities such as the CalendarCourse and used by different parts of our program. 

Design patterns and SOLID principles feature 6 follows includes factory design pattern, builder design pattern, dependency injection, and much more. 

Feature 6 also allows a saved timetable to be exported and a timetable file to be imported as well as which is a work in progress. How it would work is that with MainMenuUI, it allows Users to import a Timetable in the form of a JSON file from their local storage/computer and add it as a new Timetable in TimetableUI/AllTimetableUI and store it. The program would also allow users to export a Timetable in the form of a JSON file by grabbing the associated JSON file saved in the program’s database.

### PR DESCRIPTION
**Pull request by moonsdust**
**Summary**: 
In Feature 6, a JSON file containing all courses at the University of Toronto Faculty of Arts & Science from the Fall (F) and Winter (S) sessions can be read and converted into different entities such as the CalendarCourse and used by different parts of our program. Design patterns feature 6 follows includes builder design pattern. Feature 6 also allows a saved timetable to be exported and a timetable file to be imported as well as which is a work in progress. How it would work is that with MainMenuUI, it allows Users to import a Timetable in the form of a JSON file from their local storage/computer and add it as a new Timetable in TimetableUI/AllTimetableUI and store it. The program would also allow users to export a Timetable in the form of a JSON file by grabbing the associated JSON file saved in the program’s database. 

**Implementation Details**: 
Classes that has been implemented includes SessionGateway, SessionGatewayInteractor, SessionBuilderInteractor, GatewayInterface.etc. Currently, how it works is that using SessionGatewayInteractor, you will pass in the string of the file path to the JSON file containing all courses and then use its fileToString method to convert the JSON file into a string. From there pass this string to the method, readFromFile to create a HashMap that maps from String (Key) to CalendarCourse (Value). You can then either choose to create a single session for either the Fall (F) or Winter (S) (extractSession) or both (creatingSessionsFromFile). 

I will be further creating more classes so we can import Timetables and read them for the rest of the program to use, classes for Timetables to be exported, and create a popup UI to indicate that the Timetable has been successfully imported. I will also include Observers and Listeners soon and clean up architecture of my code.

### CODE REVIEW

File: src/main/java/fileio_use_case/CalendarCourseBuilderInteractor.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+    protected String courseCode;
+    protected String breadth;
+    public CalendarCourseBuilderInteractor(String title, ArrayList<Section> sections, String courseSession, String courseCode, String breadth) {
+        this.title = title;
+        this.sections = sections;
+        this.courseSession = courseSession;
+        this.courseCode = courseCode;
+        this.breadth = breadth;
+    }
+    public CalendarCourse newCourse(){
```
Comment (JD-Grimm) [inline src/main/java/fileio_use_case/CalendarCourseBuilderInteractor.java:22]: Is this clean?

File: src/main/java/screens/feature_6_frameworks_drivers/SessionGateway.java [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+                    (String) courseInfo.get("code"), (String) courseInfo.get("breadth")).newCourse();
+            allCourses.put((String) courseInfo.get("code"), calCourse);
+        }
+        // Return all courses in the JSON file
+        return allCourses;
+    }
+    /**
+     * Helper method of readFromFile to check for conditions of block's values and builds block
+     */
+    public Block readFromFileHelper(JSONObject aBlock) {
```
Comment (JD-Grimm) [inline src/main/java/screens/feature_6_frameworks_drivers/SessionGateway.java:77]: There is a more efficient way to do this without brute forcing through every possibility

File: src/main/java/fileio_use_case/CalendarCourseBuilderInteractor.java (hunk 1 above)
Comment (moonsdust) [reply src/main/java/fileio_use_case/CalendarCourseBuilderInteractor.java:22]: Yes, I think since interactors can work with entities

**Review by JD-Grimm — APPROVED** [review]
Overall most things look good, your file reader could be simplified, but I am not sure how important using a more elegant file reader would be for our project. Although, there is hard dependency going on in SessionGatewayInteractor, which can be fixed using an interface.

**Comment by moonsdust** [comment]
> Overall most things look good, your file reader could be simplified, but I am not sure how important using a more elegant file reader would be for our project. Although, there is hard dependency going on in SessionGatewayInteractor, which can be fixed using an interface.

I will try to see if I can fix the hard dependency and make sure my code adheres to clean architecture more. I will also try to simplify the file reader more.

**Pull request merged**